docs/training-run/index.txt

```diff
@@ -4,106 +4,255 @@
  Run Training Code
 ###################
 
-In this document, we will cover how to run your training code. You can submit your code to a cluster
-and run them as experiments. We will go over the whole cycle of it.
+This section covers how to run training code by submitting your code to a cluster and running it as
+an experiment.
 
 **********************
- Create an Experiment
+ Configure a Launcher
 **********************
 
-Currently, we only support the use of the CLI to create an experiment from scratch. You can
-alternatively create an experiment from an existing experiment or trial on the WebUI. The CLI
-command to create an experiment is as follows:
+The entrypoint of a model is configured in the :ref:`Experiment Configuration
+<experiment-configuration>` file, and specifies the path of the model and how it should be launched.
+Predefined launchers are provided in the ``determined.launch`` module and custom scripts are also
+supported.
 
-.. code::
+The launcher is configurable in the experiment configuration ``entrypoint`` field. The
+``entrypoint`` trial object or script launches the training code.
 
-   $ det experiment create <configuration file> <context directory>
+.. code:: yaml
+
+   entrypoint: python3 -m (LAUNCH_LAYER) (TRIAL_DEFINITION)
+
+or a custom script:
+
+.. code:: yaml
+
+   entrypoint: python3 script.py arg1 arg2
+
+Entrypoints for pre-configured launch layers may differ slightly in arguments, but have the same
+format:
+
+.. code:: bash
+
+   python3 -m (LAUNCH_MODULE) (--trial TRIAL)|(SCRIPT...)
+
+where ``(LAUNCH_MODULE)`` is a Determined launcher and ``(--trial TRIAL)|(SCRIPT...)`` refers to the
+training script, which can be in a simplified format that the Determined launcher recognizes or a
+custom script.
+
+Determined supports the following predefined launchers provided in the determined.launch module:
+
+-  Horovod
+-  PyTorch Distributed
+-  DeepSpeed
+
+Training Code Definition
+========================
+
+To launch a model or training script, either pass a trial class path to --trial or run a custom
+script that runs the training code. Only one of these can be used at the same time.
+
+Trial Class
+-----------
+
+.. code:: bash
+
+   --trial TRIAL
+
+To specify a trial class to be trained, the launcher accepts a TRIAL argument in the following
+format:
+
+.. code:: yaml
+
+   filepath:ClassName
+
+where filepath is the location of your training class file, and ClassName is the name of the Python
+training class
+
+Custom Script
```

Comment:
Since a user can submit a bare script without a launcher, I'm not sure quite where this comment belongs.

But if a dtrain user writes their own launcher, it needs to have the following properties:

- Launch all desired workers, passing any required peer info (like rank or chief ip address) to each worker
- Monitor workers.  If any worker exits nonzero, the launcher should kill the remaining workers and exit nonzero.
- Exit nonzero after all workers exit nonzero

Those requirements ensure that distributed training jobs don't hang after a single worker failure.

Reply:
edit to add that a custom launcher SHOULD wrap each rank worker in the 
```
python3 -m determined.launch.wrap_rank $RANK CMD [ARGS...]
```
script, so that in the webui the final logs can be separated according to rank.  This is also described in the Core API distributed training example.

Reply:
yeah, i wasn't quite sure where this belonged either. there's two levels of "custom" in the context of entrypoint: custom launcher, and custom script. perhaps this belongs in the entrypoint doc

Reply:
Not resolved. Please, review where I put it in the latest doc update.

Reply:
I think it's good where it's at.